### ISSUE TITLE
My Projects List > Add ability to "save" or "fave" my own projects

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#1629]
I think it would be a good feature to be able to "star" or "fav" my own projects. They would appear at the top of the list always. This could alleviate this feeling of things being hard to find within the list, which I feel personally, and that we've heard about from users.

**Comment by kramsamujh** [comment]
I'm interested in working on this bug.

**Comment by flukeout** [comment]
Sounds good, @kramsamujh - I think what we'd need done is to have an extra column in the projects database that would keep track of this property, then we can sort the list with those items at the top. 

Also, if someone clicks on an item to make it their favorite, we should move it to the top of the list immediately.

@gideonthomas can you weigh in on this? Would it be a big lift to add an additional database column in the projects table? Is there a lighter-weight way you can think of doing this?

**Comment by gideonthomas** [comment]
@flukeout this is not a good first bug if we want full functionality reliant on a backing database because it will be blocked on a couple of changes to the `publish.webmaker.org` db (not just adding another column to the db but also adding routes to edit values in that column).

One alternative would be to instead store favorites in `localStorage` (though I feel like it might look clunky because we would need to alter the position of DOM elements using JS). It's still a little bit of work but is significantly less compared to relying on the publish database.

**Comment by Pomax** [comment]
Keep it a good first bug, but let's restrict it to localStorage for now. If that works, syncing to a db is a perfectly manageable follow-up.

**Comment by flukeout** [comment]
I'm happy with giving the localStorage solution a go. We'll need to rebuild the list every time something is marked / unmarked as a favorite.

**Comment by humphd** [comment]
+1 to using `localStorage` to start.  @kramsamujh if you're not already familiar with it, see https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API/Using_the_Web_Storage_API.

**Comment by gideonthomas** [comment]
Here are some more details about what you would need to implement this:
1. **Allowing favoriting of projects**
   - We probably want to store the data about favorites as a javascript object in `localStorage`. Let's say the `localStorage` key is `project-favorites`, its value will be an array, for e.g. `[ 1051, 1113, 1551 ]`, where each of those numbers are project ids. You will probably need to use [JSON.parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse) and [JSON.stringify](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) to get/store arrays in `localStorage` since `localStorage` usually stores only strings.
   - Depending on where the favorite button would go, you would add HTML somewhere around [here](https://github.com/mozilla/thimble.mozilla.org/blob/master/views/editor/projects.html#L29-L40) to add another `<td>` element that will contain your "favorite button" in the unfavorited state. You can add any CSS style rules for the element you create, in [this file](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/stylesheets/projects-style.less)
      - @flukeout are you able to provide some design direction here (position of the favorite button + either the asset or guidance on writing HTML/CSS to create a favorite button)?
   - Somewhere around [here](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L21-L26) create a variable that will get the list of favorites from `localStorage`.
   - Inside [this forEach function](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L33-L41), use a query selector (similar to what's already done in that function) to get the "favorite button" element for that project row and add a `click` event listener (similar to [this code](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L37-L39)). Inside the event listener, get the array from `localStorage` again and if the project id (which you have access to [because of this](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L34)) is in the array, remove it from the array, otherwise, add it to the array and save it back to `localStorage`. Also, depending on how you designed it, mark the element as "favorited" (or "unfavorited") (this could be as simple as adding a css class to it if the favorite button was purely HTML). I also suggest creating a variable with an empty array outside the forEach function and storing "project"([from here](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L33)) DOM elements in that array if it is a favorite project.
   - Inside the same forEach function, check if the project id is in the array. If so, get the "favorite button" element for that project row and mark it as "favorited"
2. **Moving the favorite projects to the top of the list**
   - Let's try to keep this simple for now and not worry about performance/looks for the moment and let's only show it at the top of the list when we load the page (that is to say, we won't worry yet about where the project row "moves" when you hit the favorite button). We can make these improvements iteratively.
   - Out [here](https://github.com/mozilla/thimble.mozilla.org/blob/master/public/editor/scripts/project-list.js#L42), use the array you created in the previous step containing the project DOM elements that were favorites and pass it as an argument into the jquery method ["prepend"][http://api.jquery.com/prepend/] called on `$("#project-list")`. So it will look something like `$("#project-list").prepend(arrayContainingFavoriteProjectDomElements)`

@kramsamujh you should probably start with step 1 and see how far you come along. If you need any help (or find the instructions confusing), feel free to reach out to us here via comments.

**Comment by gideonthomas** [comment]
@kramsamujh how are you doing with this bug?

**Comment by flukeout** [comment]
For future reference, the work here has been continuing in https://github.com/mozilla/thimble.mozilla.org/pull/1731

**Comment by humphd** [comment]
https://github.com/mozilla/thimble.mozilla.org/pull/1731 was merged, can we close this bug @flukeout?

**Comment by flukeout** [comment]
Definitely!
